tests/unit/test_baseObject.py

```diff
@@ -109,3 +110,63 @@ def test_decoratedScriptsAndGestureDictionariesIfDynamic(self):
 		obj = DynamicNVDAObjectWithDecoratedScriptAndGesturesDictionary()
 		for key in ("a", "b", "c", "d", "g", "h"):
 			self.assertIn("kb:%s" % key, obj._gestureMap)
+
+class AutoPropertyObjectWithAbstractProperty(AutoPropertyObject):
+
+	_abstract_x = True
+	def _get_x(self):
+		return True
+
+	def _set_x(self, value):

```

Comment:
what about abstract set? If set and get need different patterns, perhaps `_abstract_get_x` and `_abstract_set_x`

Reply:
In the current implementation, you mark an entire property as abstract. Note that *not the methods*, but the property itself is considered abstract, which is also the case when using abc.abstractproperty.

Reply:
This is clearer to me now thanks to your other explanation. However, I'm concerned that its not so obvious from the syntax used to mark the property as abstract. Essentially what `_abstract_x` is saying is "use magic to declare that this class must have a property called `x` but its implementation is left to subclasses". It makes me wonder though.. is there any point in declaring `_get_x` and `_set_x` methods. This makes `x` look very much NOT abstract.

Reply:
This looks odd, but is behavior defined by Python. It might be helpful to read [the docs for the Abstract module](https://docs.python.org/2/library/abc.html#id1)

In short: an abstract method or or property doesn't mean that the method or property should be left unimplemented, since a sub class can call the method or get the property of the abstract base class.

Also note from the Python docs:

> C++ programmers should note that Python’s virtual base class concept is not the same as C++’s.

Does this help?

Reply:
>In short: an abstract method or or property doesn't mean that the method or property should be left unimplemented, since a sub-class can call the method or get the property of the abstract base class.

But it doesn't **have** to be implemented in the base class, right? Except there is the note in the AutoPropertyObject about set overriding the property. 

> C++ programmers should note that Python’s virtual base class concept is not the same as C++’s.

Hahaha this would help a lot more if the differences were described.

I think I must be missing the point when it comes to properties. I assume the goal is: allow a base class to declare that a certain property must exist in the construction of a subclass, without necessarily providing the implementation for that property in the base class?

Reply:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [1 aug. 2018 11:19 CEST](https://github.com/nvaccess/nvda/pull/8393#discussion_r206809374 "2018-08-01T09:19:29Z - Replied by Github Reply Comments"):
> But it doesn't **have** to be implemented in the base class, right? Except there is the note in the AutoPropertyObject about set overriding the property.

No, it doesn't have to. In that case the property getter can raise NotImplementedError as we already do.

> Hahaha this would help a lot more if the differences were described.

I agree. I recall I've seen them somewhere, but couldn't find them that quickly.


> I think I must be missing the point when it comes to properties. I assume the goal is: allow a base class to declare that a certain property must exist in the construction of a subclass, without necessarily providing the implementation for that property in the base class?

Yes, you're correct. The bookmark property on textInfos.TextInfo is probably a good example of that.

Reply:
> I agree. I recall I've seen them somewhere, but couldn't find them that quickly.

I couldn't find a good explanation either.

>Yes, you're correct. The bookmark property on textInfos.TextInfo is probably a good example of that.

Ok, then my concern is really about what "implemented" means. Since, there are several options:
- A base class says a property must exist, it can be read-only (set raises `NotImplementedException`) read/write or write-only? The base class does not care which as long as something is implemented in the sub-class. I suppose this is the use case for `_abstract_x`
- A base class says a property exists, but some part of it MUST be implemented by the sub-class. For instance the base class might know how to provide the get implementation, but not the set. The _set_x could be marked as @abstractmethod in this case?

Reply:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [6 aug. 2018 11:08 CEST](https://github.com/nvaccess/nvda/pull/8393#discussion_r207821232 "2018-08-06T09:08:28Z - Replied by Github Reply Comments"):
> Ok, then my concern is really about what "implemented" means. Since, there are several options:
> - A base class says a property must exist, it can be read-only (set raises `NotImplementedException`) read/write or write-only? The base class does not care which as long as something is implemented in the sub-class. I suppose this is the use case for `_abstract_x`

Note that within AutoPropertyObject, there is no such thing as a read only property. When someone only implements _get_x, setting x will override the property completely.

> - A base class says a property exists, but some part of it MUST be implemented by the sub-class. For instance the base class might know how to provide the get implementation, but not the set. The _set_x could be marked as @abstractmethod in this case?

In the current implementation, the whole property would be marked as abstract. As soon as the setter is defined in the sub class,  the abstractness of the property goes away, unless _abstract_x is also explicitly set to True for the sub class.

Either decorating _get_x or _set_x with abstractmethod is dangerous as I noted above, as it marks the method as abstract, not the property.

```
class Test(AutoPropertyObject):

	@abstractmethod
	def _get_x(self):
		raise NotImplementedError

class Subtest(Test):
	x = "This won't initialize, as _get_x is still abstract.
```

I think I can live with it if we decide to do the abstraction at the method level, it might look easier to understand. However in that case, our implementation of abstract properties doesn't match the pythonic application of it, and that is what worries me.

Reply:
> Note that within AutoPropertyObject, there is no such thing as a read only property. When someone only implements _get_x, setting x will override the property completely.

It would be nice if we could find a way to catch the situation where _get_x is provided but not _set_x. Is there is a situation in which this behaviour is desirable?

> Either decorating _get_x or _set_x with abstractmethod is dangerous as I noted above

But if it's not abstract, then the base class should be providing it:
```
class Test(AutoPropertyObject):

	def _get_x(self):
		return 42

	@abstractmethod
	def _set_x(self, val):
		raise NotImplementedError

class Subtest(Test):
	# Subtest shouldn't be able to instantiate because _set_x is abstract.
	x = 84
```

> I think I can live with it if we decide to do the abstraction at the method level, it might look easier to understand.

I'm not necessarily arguing for that, but I do want to make sure I understand why it's done this way, and what the alternatives might be. I like the `@abstractproperty` syntax in the [examples on the `abc` module docs](https://docs.python.org/2/library/abc.html#abc.abstractproperty)

Reply:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [7 aug. 2018 10:57 CEST](https://github.com/nvaccess/nvda/pull/8393#discussion_r208149934 "2018-08-07T08:57:31Z - Replied by Github Reply Comments"):
> > Note that within AutoPropertyObject, there is no such thing as a read only property. When someone only implements _get_x, setting x will override the property completely.
> 
> It would be nice if we could find a way to catch the situation where _get_x is provided but not _set_x. Is there is a situation in which this behaviour is desirable?

There are actually quite a lot of cases where _get_x is declared and _set_x is not. See displayModel.DisplayModelTextInfo, _get_backgroundSelectionColor and _get_foregroundSelectionColor. These cases perfectly explain the use of this functionality, namely fetching a value the first time the property is retrieved.

What do you actually want to catch exactly? I'd rather catch cases where _set_ is implemented  and _get_ isn't, because those make less sense to me.

> > > Either decorating _get_x or _set_x with abstractmethod is dangerous as I noted above
> 
> But if it's not abstract, then the base class should be providing it:
> ```
> class Test(AutoPropertyObject):
> ...
> ```

Your example looks nice, but I"m afraid that it doesn't work as you expect it to work, or I"m just not getting what you're trying to point out.

```
class Test(AutoPropertyObject):

	def _get_x(self):
		return 42

	def _set_x(self, val):
		print("Magic setter")

class Subtest(Test):
	x = 84

test = Test()
print(test.x)
42
test.x = "random"
"Magic setter"
print(test.x)
42

subtest = Subtest()
# Note that setting x on the subclass wil override x completely, so it will abandon both the getter and the setter. Therefore, overriding x when it is marked as abstract should make it no longer abstract?.
print(subtest.x)
84
# Note that this does no longer print anything.
subtest.x = "random"
print(subtest.x)
84
```

> > I think I can live with it if we decide to do the abstraction at the method level, it might look easier to understand.

Thinking about the example I gave above, I guess that doing abstraction at the method level will also cause confusion anyway..

Reply:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [8 aug. 2018 04:20 CEST](https://github.com/nvaccess/nvda/pull/8393#discussion_r208438777 "2018-08-08T02:20:43Z - Replied by Github Reply Comments"):
> Are you aware of a reason for this behaviour? 

Yes, see this code snipped from baseObject. 

```
			if not g:
				# There's a setter or deleter, but no getter.
				# This means it could be in one of the base classes.
				for base in bases:
					g = getattr(base,'_get_%s'%x,None)
					if g:
						break
```

It seems to break the polymorphic behaviour that I would expect. I suppose it could be that I'm tying the interface of the property to the interface of the class, and instead the interface of the property (and it's type) should not be considered part of the class that holds it, merely the existence of a property called x

Reply:
@feerrenrut: Do you want me to change anything in this behaviour, or do you have any other concerns?

Reply:
No, I think it's ok for now.